### ISSUE TITLE
FindPETSc fails to find PETSc if installed version not in pkg-config search path

### ISSUE THREAD

**Issue by ajaust** [precice/precice#888]
**Describe your setup**

Operating system (e.g. Linux distribution and version): _Ubuntu 18.04_ (or Debian 10.6)
preCICE Version: develop

On Ubuntu I have PETSc `3.7.7` installed via `apt` and a newer version `3.13.3` compiled by hand for which I set the environment variables via modules. The following environment variables are set in case `3.13.3` ist used

```
local base = "/home/jaustar/software/petsc/3.13.3-opt"
local arch = "arch-linux-c-opt"

setenv(       "PETSC_DIR", base )
setenv(       "PETSC_ARCH", arch )
prepend_path( "LD_LIBRARY_PATH", pathJoin( base,"lib" ) )
prepend_path( "CPATH", pathJoin( base, "include" ) )
prepend_path( "LIBRARY_PATH", pathJoin( base, "lib" ) )
prepend_path( "PYTHONPATH", pathJoin( base, "lib" ) )
```

In my case preCICE resides in `${PETSC_DIR}`! However, `${PETSC_DIR}/${PETSC_ARCH}` could be legal, tool.

**Describe the problem**

The new `FindPETSc.cmake` script does fails to find PETSc if PETSc is not known to `pkg-config`, i.e. if `PKG_CONFIG_PATH` does not contain the PETSc installation path (or rather the correspondig subdirectory).

**Step To Reproduce**
1. Install PETSc in a directory that is not in the `pkg-config` search path.
2. Set `PETSC_DIR` and `PETSC_ARCH`
3. Try to compile preCICE
4a. On Ubuntu PETSc 3.7.7 was detected and the compilation process was aborted due to old version.
4b. On Debian the compilation failed because no PETSc installation is found
```
Could NOT find PETSc (missing: PETSc_FOUND PETSc_INCLUDE_DIRS) (Required is
  at least version "3.12")
```

After extending `PKG_CONFIG_PATH` to point to `${PETSC_DIR}/lib/pkgconfig/` or `${PETSC_DIR}/${PETSC_ARCH}lib/pkgconfig/` (depending on the actual instllation path) preCICE was able to detect PETSc correctly.

**Expected behaviour**

preCICE detects the PETSc installation defined by `PETSC_DIR` and `PETSC_ARCH` and compiles correctly. At least that was the behavior with the old `FindPETSc.cmake`. 

Maybe the new script should try harder to find PETSc in the given `$PETSC_DIR` or `${PETSC_DIR}/${PETSC_ARCH}` before looking other places.

**Additional context**

The old `FindPETSc.cmake` worked with only `PETSC_DIR` and `PETSC_ARCH` being set.

I observed the problem myself and it was reported by user guifon1000 on Gitter. The user used Debian 10.6 and had PETSc installed in `${PETSC_DIR}/${PETSC_ARCH}`.

**Comment by MakisH** [comment]
Then one could argue that this is a kind-off breaking change? In any case, if we release the new version, we need to remember to communicate this in the release notes.

**Comment by ajaust** [comment]
I think one would have to figure out if the changed behavior was intended. If so, mention in the release notes. If not, one should adapt the script. 

In the latter case one could maybe adapt the new `FindPETSc.cmake` script to add `${PETSC_DIR}/lib/pkgconfig/` and `${PETSC_DIR}/${PETSC_ARCH}lib/pkgconfig/ ` to `PKG_CONFIG_PATH` before searching for PETSc. This should restore the old behavior. I am not sure about potential problems arising from that.

**Comment by fsimonis** [comment]
@ajaust This should work out of the box.

It looks like the module does not include the `CMAKE_PREFIX_PATH`. 

Can you please check if the modules properly detects PETSc after setting the CMake variable `PKG_CONFIG_USE_CMAKE_PREFIX_PATH`?

**Comment by fsimonis** [comment]
To elaborate on this issue:

The CMake `pkg-config` wrapper can include the content of `CMAKE_PEFIX_PATH` to the pkg-config search paths.
I hijack this in the FindPETSc Module to inject the ENV-based paths  `$PETSC_DIR/$PETSC_ARCH` or simply `PETSC_DIR` if arch is missing.

So if the FindPETSc module cannot find PETSc even though both ENV are present, then this is a bug in the module.


@ajaust can you please add the following line at the top of the `FindPETSc.cmake` file:
```
variable_watch(CMAKE_PREFIX_PATH)
``` 
This should give you an insight on when the prefix path gets updated. My guess is that it is not extended correctly.

**Comment by ajaust** [comment]
Adding `variable_watch(CMAKE_PREFIX_PATH)` gives the following output when trying to compile `develop`:

```
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.1") 
CMake Debug Log at cmake/modules/FindPETSc.cmake:74 (if):
  Variable "CMAKE_PREFIX_PATH" was accessed using UNKNOWN_DEFINED_ACCESS with
  value "".
Call Stack (most recent call first):
  CMakeLists.txt:216 (find_package)


-- Checking for module 'PETSc>=3.12'
CMake Debug Log at /usr/share/cmake-3.10/Modules/FindPkgConfig.cmake:167 (if):
  Variable "CMAKE_PREFIX_PATH" was accessed using UNKNOWN_READ_ACCESS with
  value "".
Call Stack (most recent call first):
  /usr/share/cmake-3.10/Modules/FindPkgConfig.cmake:280 (_pkgconfig_add_extra_path)
  /usr/share/cmake-3.10/Modules/FindPkgConfig.cmake:593 (_pkg_check_modules_internal)
  cmake/modules/FindPETSc.cmake:93 (pkg_check_modules)
  CMakeLists.txt:216 (find_package)


--   Requested 'PETSc >= 3.12' but version of PETSc is 3.7.7
CMake Debug Log at cmake/modules/FindPETSc.cmake:98 (unset):
  Variable "CMAKE_PREFIX_PATH" was accessed using REMOVED_ACCESS with value
  "".
Call Stack (most recent call first):
  CMakeLists.txt:216 (find_package)


CMake Error at /usr/share/cmake-3.10/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find PETSc (missing: PETSc_FOUND PETSc_INCLUDE_DIRS) (Required is
  at least version "3.12")
Call Stack (most recent call first):
  /usr/share/cmake-3.10/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
  cmake/modules/FindPETSc.cmake:173 (find_package_handle_standard_args)
  CMakeLists.txt:216 (find_package)
```
I only copied the part of the output that seems to be related to the error to keep it short.


There is no reference to PETSc in the CMake log files.

**Comment by fsimonis** [comment]
@ajaust I extended the find module to handle more paths and providing more informative log output.
Could you try the following version of the file please  https://github.com/precice/precice/blob/fix-888/cmake/modules/FindPETSc.cmake .

It now tells you what prefixes were detected and ignored as well as the _full_ range of paths passed to pkg-config.

There are also 2 implementations of pkg-config around called `pkg-config` and `pkgconf`. One has issues handling multiple versions, the other doesn't.

**Comment by ajaust** [comment]
I am on Ubuntu 18.04 still that uses `pkg-config` as default and that is the one where I observe the problems. In my Spack environments this might be different. I think I use `pkgconf` there by default, but I am quite convinced that Spack also sets all kind of environment variables.

The new version seems to work nicely:
```
-- The CXX compiler identification is GNU 7.5.0
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test CXX11_N2071_N2072
-- Performing Test CXX11_N2071_N2072 - Success
CMake Warning at CMakeLists.txt:49 (message):
  You did not specify a CMAKE_BUILD_TYPE.

  

    We will assume you asked for a Debug build.


-- === CONFIGURATION ===
-- Library version to build: 2.1.1
-- Build configuration: Debug
-- Build shared libraries: ON
-- Target system: Linux-4.15.0-122-generic
-- Host system: Linux-4.15.0-122-generic
-- Library architecture: x86_64-linux-gnu
-- CXX compiler: /usr/bin/c++
-- CXX compiler flags: <empty>
-- CXX linker: /usr/bin/ld
-- Install prefix: /usr/local
-- Source directory: /home/jaustar/software/compilescripts/precice/precice-develop
-- Binary directory: /home/jaustar/software/compilescripts/precice/precice-develop/build_cmake_test
-- Configure package generation: ON
-- Install tests/testfiles: OFF
-- Additional CTest MPI Flags: <empty>

-- === ENABLED FEATURES ===
 * MPICommunication, Enables the MPI communication back-end.

   This enables the MPI communication back-end which is highly recommended on multi-node systems.
   See the documentation of the CMake module FindMPI to control its functionality.

   This feature can be enabled/disabled by setting the PRECICE_MPICommunication CMake option.

 * PETScMapping, Enables the PETSc-powered radial basic function mappings.

   This enables the PETSc-based variant of radial basis function mappings which can run in parallel,
   also across different compute nodes. This is highly recommended for large cases running in parallel.

   This feature can be enabled/disabled by setting the PRECICE_PETScMapping CMake option.
   Requires MPICommunication.
  
 * PythonActions, Enables the support for user-defined python actions.

   preCICE allows to manipulate coupling data at runtime using configurable actions.
   This feature enables the support for user-defined actions written in Python based on numpy.

   This feature can be enabled/disabled by setting the PRECICE_PythonActions CMake option.
  
 * CBindings, Enables the native C bindings.

   preCICE provides native bindings for C, which are compiled into the library.
   This feature enables the compilation and installation of the bindings into the library.
   Note that we strongly recommend to compile with C bindings enabled for compatibility reasons.

   This feature can be enabled/disabled by setting the PRECICE_ENABLE_C CMake option.
  
 * FortranBindings, Enables the native Fortran bindings.

   preCICE provides native bindings for Fortran, which are compiled into the library.
   This feature enables the compilation and installation of the bindings into the library.
   Note that we strongly recommend to compile with Fortran bindings enabled for compatibility reasons.

   This feature can be enabled/disabled by setting the PRECICE_ENABLE_FORTRAN CMake option.
  

-- === DEPENDENCIES ===
-- Looking for C++ include pthread.h
-- Looking for C++ include pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE  
-- Boost version: 1.65.1
-- Found the following Boost libraries:
--   filesystem
--   log
--   log_setup
--   program_options
--   system
--   thread
--   unit_test_framework
--   date_time
--   regex
--   chrono
--   atomic
-- Found Eigen3: /usr/include/eigen3 (Required is at least version "3.2") 
-- Validating Eigen
-- Validating Eigen - success
-- Found LibXml2: /usr/lib/x86_64-linux-gnu/libxml2.so (found version "2.9.4") 
-- Validating LibXml2
-- Validating LibXml2 - success
-- Validating JSON
-- Validating JSON - success
-- Validating Prettyprint
-- Validating Prettyprint - success
-- Found MPI_CXX: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi_cxx.so (found version "3.1") 
-- Found MPI: TRUE (found version "3.1")  
-- MPI Version: Open MPI v2.1.1, package: Open MPI buildd@lcy01-amd64-009 Distribution, ident: 2.1.1, repo rev: v2.1.0-100-ga2fdb5b, May 10, 2017
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.29.1") 
-- Detecting additional PETSc prefixes
-- Detected /home/jaustar/software/petsc/3.13.3-opt
-- Skipped the following invalid prefixes: /home/jaustar/software/petsc/3.13.3-opt/arch-linux-c-opt
-- pkg-config will search the following paths:
CMAKE_PREFIX_PATH
   /home/jaustar/software/petsc/3.13.3-opt
-- Checking for module 'PETSc>=3.12'
--   Found PETSc, version 3.13.3
-- Found PETSc: TRUE (found suitable version "3.13.3", minimum required is "3.12") 
-- Found PythonLibs: /usr/lib/x86_64-linux-gnu/libpython3.6m.so (found version "3.6.9") 
-- Found PythonInterp: /usr/bin/python3.6 (found version "3.6.9") 
-- Found NumPy: /home/jaustar/.local/lib/python3.6/site-packages/numpy/core/include (found version "1.16.3") 
-- Validating LibPython
-- Validating LibPython - success
-- Validating NumPy
-- Validating NumPy - success

-- === TARGETS & PACKAGES ===
-- Including test sources
-- Compressing manpage: /home/jaustar/software/compilescripts/precice/precice-develop/build_cmake_test/packaging/man1/binprecice.1
Debian package generator disabled: Install prefix is not "/usr"
-- Compressing changelog

-- === TESTS ===
-- Looking for a C compiler
-- Looking for a C compiler - /usr/bin/cc
-- Looking for a Fortran compiler
-- Looking for a Fortran compiler - /usr/bin/f95
-- Test precice.acceleration
-- Test precice.action
-- Test precice.com
-- Test precice.com.mpiports - skipped (OpenMPI)
-- Test precice.cplscheme
-- Test precice.io
-- Test precice.m2n
-- Test precice.m2n.mpiports - skipped (OpenMPI)
-- Test precice.mapping
-- Test precice.mapping.petrbf
-- Test precice.math
-- Test precice.mesh
-- Test precice.partition
-- Test precice.interface
-- Test precice.serial
-- Test precice.parallel
-- Test precice.query
-- Test precice.testing
-- Test precice.utils
-- Test precice.xml
-- Test precice.solverdummy.build.cpp
-- Test precice.solverdummy.build.c
-- Test precice.solverdummy.build.fortran
-- Test precice.solverdummy.run.cpp-cpp
-- Test precice.solverdummy.run.c-c
-- Test precice.solverdummy.run.fortran-fortran
-- Test precice.solverdummy.run.cpp-c
-- Test precice.solverdummy.run.cpp-fortran
-- Test precice.solverdummy.run.c-fortran
-- Configuring done
-- Generating done
-- Build files have been written to: /home/jaustar/software/compilescripts/precice/precice-develop/build_cmake_test
```

Compilation went through nicely as well and `make test` finished without any errors.

**Comment by fsimonis** [comment]
> PETSc is installed in ${PETSC_DIR}

Oh I see!
When you install PETSc to a prefix, then the `PETSC_ARCH` is merged with the content of `PETSC_DIR`.
My FindPETSc was not able to handle that case.

**Comment by ajaust** [comment]
I considered that a legal case (so far). I basically copy the values for `PETSC_DIR` and `PETSC_ARCH` from the PETSc install script which still gave me a value for `PETSC_ARCH` the last time I checked even though it was not part of the installation prefix.
